### ISSUE TITLE
XSLT Help

### ISSUE THREAD

**Issue by nlottig94** [nlottig94/EmilyDickinson6#11]
@ebeshero 
This is the [XSLT from Dickinson 16](https://github.com/ebeshero/EmilyDickinson16/blob/master/DickinsonLineNumbers/dickinson.xsl) that Alex made, and it works for Fascicle 16. [I have added and deleted things that were needed/not needed.](https://github.com/nlottig94/EmilyDickinson6/blob/master/Dickinson6_InitialXML/dickinson.xsl)  However, for some odd reason, when an HTML is output, it doesn't work on Fascicle 6 (particularly on [Poem 4](http://htmlpreview.github.io/?https://github.com/nlottig94/EmilyDickinson6/blob/master/Web-Files/6/604.html)).  My issue here is that the XSLT isn't picking the pieces of the lines out in the correct order and is also not outputting spaces between `app`s.  I also tried to create tables in the output ([XSLT here](https://github.com/nlottig94/EmilyDickinson6/blob/master/Web-Files/6/xml/Dickinson6Experiment2.xsl)), but it's not working either ([HTML output here](http://htmlpreview.github.io/?https://github.com/nlottig94/EmilyDickinson6/blob/master/Web-Files/6/604Experiment2.html)).

**Comment by nlottig94** [comment]
@ebeshero 
You can also look at all of the others [here](https://github.com/nlottig94/EmilyDickinson6/tree/master/Web-Files/6).  [Poem 1](http://htmlpreview.github.io/?https://github.com/nlottig94/EmilyDickinson6/blob/master/Web-Files/6/601.html) looks okay, but as I was looking through the pages, Poem 4 is just really messed up :(

**Comment by ebeshero** [comment]
@nlottig94 I'm looking at the XSLT that creates table output first. The output error you're getting is due to the `<span>` element, which is an inline HTML element and isn't allowed to wrap around a block element like `<table>`. So you can't wrap `<table>` or any table element in a `<span>`. It's okay to have the `<span>` elements be sitting _inside_ table cells (`<td>`) though. You'll need to rethink your strategy here. 

There's another problem with the tables that I can see, and it's that you're outputting a lot of empty table cells. I'm not sure why that is, so I'll take a closer look now...

See http://stackoverflow.com/questions/11314668/what-elements-can-a-span-tag-contain-in-html5

**Comment by ebeshero** [comment]
@nlottig94 Aha! The answer to why you're getting so many extra output `<td>` cells is simple! In your XSLT, you have put the `<td>` elements _outside_ your `<xsl:if>` conditional statements, and that means you are _always_ outputting them, but they are coming out empty when your condition test isn't being met. 

Try fixing these things (and you'll probably have to make a special modified CSS and JavaScript, too)...and let's see how your output comes out! (This is just the XSLT outputting tables, now).

**Comment by ebeshero** [comment]
@nlottig94 In _both_ XSLT files (the one making a table and the one outputting your original format) there's the same problem, isn't there--that you're outputting the contents of the lines in the wrong order. Let's see if we can figure out why that's happening with the processing order.

**Comment by HelenaSabel** [comment]
Hello! I modified an XSLT file called "Dickinson6Experiment4.xsl" because I thought it was the most recent version of this bug. I only changed the part where the tables were being created so I didn’t touch anything before that part. 
I think the tables are correctly created with the format I understood you wanted:
- One table for each line
- One row for each witness
- One cell for each node

Let me know if it doesn’t work or if that wasn’t the output you wanted!

**Comment by nlottig94** [comment]
@HelenaSabel thank you so much for your help!!
Okay, so your XSLT wasn't quite what I wanted to do, so [I messed around a bit with it](https://github.com/nlottig94/EmilyDickinson6/blob/master/Web-Files/6/xml/Dickinson6Experiment4.xsl)....and I'm getting [the output that I want](http://htmlpreview.github.io/?https://github.com/nlottig94/EmilyDickinson6/blob/master/Web-Files/6/604.html)!!  The one thing that doesn't seem to be working right is my CSS.  I have a `div` with an `@class`. The way the [CSS](https://github.com/nlottig94/EmilyDickinson6/blob/master/Web-Files/css/styles.css) is set up, it _should_ be outputting highlight color...Could someone tell me why it's not working?

**Comment by nlottig94** [comment]
@HelenaSabel @ebeshero @brookestewart @RJP43 
Hi again! So I've figured out my coloring issue, but I'm having a different coloring issue.
Here is the [XSLT](https://github.com/nlottig94/EmilyDickinson6/blob/master/Web-Files/6/xml/Dickinson6Experiment4.xsl).
And here is the [output](http://htmlpreview.github.io/?https://github.com/nlottig94/EmilyDickinson6/blob/master/Web-Files/6/604-1.html).
I am coloring and outputting each variant in a separate table line, which is awesome, and it's being colored!!  However, I can't figure out how to only color the stuff in the `td` cells that hold the `rdg` elements...Any ideas??

**Comment by ebeshero** [comment]
@nlottig94 @HelenaSabel First, thanks Helena for helping us out with _exactly_ what we were stuck on last night: We weren't sure how to store information from the template match context on the `<l>` element in the new context set in the `<xsl:for-each>`, and my solution was `<xsl:for-each-group>`, but yours with **storing the `current()` template match info** inside a variable and calling it inside the for-each loop is exactly what we needed!  And I find we can simplify it just a little. You don't actually need to use `<xsl:sequence>` in defining that $current variable: You don't really need:

```
  <xsl:variable name="current">
            <xsl:sequence select="current()"/>
        </xsl:variable>    
```

because this simply seems to work as the template fires on its own matching on the line elements:

```
<xsl:variable name="current" select="current()"/>
```

That's just enough.

I was really glad to learn about `<xsl:param>` and `<xsl:with-param>` which I've never used before. The `@tunnel` attribute is wonderfully helpful since it passes a value to all other templates called after the current one! I'm curled up right now reading about in Michael Kay on pp. 425+ and Nicole, you'll want to read about this. If I understand this right, the `<xsl:param>` definition with an `@tunnel` applied to it will make it possible for you to read a `@wit` attribute down inside an `<app>` element and make it available for use when it's called elsewhere in the XSLT stylesheet, inside other template matches. It's a neat trick--and again, just the sort of thing to help us grab what we need from one level of template matches and call it when we want it at a different level. (Am I understanding that right, Helena?)

**Comment by ebeshero** [comment]
@nlottig94 Armed with some information about how `<xsl:param>` and `<xsl:with-param>` work together with that super-cool `@tunnel` in different template rules, I bet we can figure out how to position your `@class` attribute on only the table data cells that contain output from the `<app>` element when it outputs a designated `<rdg>` (based on its `@wit`)...

**Comment by ebeshero** [comment]
@nlottig94 @HelenaSabel I figured it out! (Thanks again, Helena, for showing us xsl:param!) Nicole, yes, you can use the param tunnel to deliver the value! I'll push my code in a minute. 
@RJP43 @spadafour I think you'll both find this interesting... :+1:

**Comment by ebeshero** [comment]
Here ya go! @nlottig94 
https://github.com/nlottig94/EmilyDickinson6/pull/14

**Comment by nlottig94** [comment]
Thanks @ebeshero!  And thanks @HelenaSabel!!  I'll merge the pull request in the morning.  I'm heading to bed early because of the lack of sleep I got last night!  Thank you so so much!!

**Comment by nlottig94** [comment]
@ebeshero @HelenaSabel
Thanks so much for all your help! I'm now trying to figure out how to hide the _entire_ table row when the publications are switched off. Right now (in @ebeshero's output) when you toggle the publication off, it still leaves the non-`rdg` text on.

**Comment by HelenaSabel** [comment]
Hi again! I’m glad  I could be of help although I misunderstood what you wanted to do. I thought you wanted a table where you can simultaneously read all the witness (and their variants) at the same time. That’s why the common text is repeated in each row. However, if what you want to do is to toggle between readings, it makes much more sense (and a much cleaner output) to have the different variants inside `<span>` elements and only toggle those.

**Comment by nlottig94** [comment]
@ebeshero @HelenaSabel!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
I'VE FIGURED IT OUT WITH JAVASCRIPT AND CSS!!!!!!!
Give me a few minutes to work out the bugs, and then I'll post something!!!!

**Comment by nlottig94** [comment]
@ebeshero @HelenaSabel @brookestewart 
Okay!
Here is the finished [XSLT](https://github.com/nlottig94/EmilyDickinson6/blob/master/Web-Files/6/xml/Dickinson6FINAL.xsl).
Here is the changed [CSS](https://github.com/nlottig94/EmilyDickinson6/blob/master/Web-Files/css/styles.css).
Here is the changed [JavaScript](https://github.com/nlottig94/EmilyDickinson6/blob/master/Web-Files/scripts/dickinson.js).
And here is the [final output of Poem 4](http://htmlpreview.github.io/?https://github.com/nlottig94/EmilyDickinson6/blob/master/Web-Files/6/604.html)!!

I don't like the look that the table cells give, so I'm leaving it like this!!  What do you all think????

**Comment by HelenaSabel** [comment]
Very nice job! 
I have a few questions, though:
- Why is the pronoun “I” highlighted in line 6?
- Are the white spaces before punctuation variants generated on purpose or are those the result of handling the `app/rdg` elements?

**Comment by nlottig94** [comment]
@HelenaSabel 
First, I is highlighted because in the different versions, have it underlined, italicized, or nothing.  You can see it now!!!
Next, the white spaces before the punctuation is because of the table cells...

**Comment by ebeshero** [comment]
@nlottig94 Nicole: Have you collapsed the borders of those cells? Here's the CSS property you want:
http://www.w3schools.com/cssref/pr_border-collapse.asp

Basically you want the table cells to be as minimal as possible...

**Comment by nlottig94** [comment]
Thanks @ebeshero and @HelenaSabel for all your help!

**Comment by nlottig94** [comment]
Okay @ebeshero and @HelenaSabel....I know that you don't want to hear this, but I have a new issue..
So here is the background: After changing what needed to be changed in the XSLT to make the poems in Fascicle 16 look like the poems in Fascicle 6 (all in tables, etc.), I ran the XSLT over all the poems and I thought it looked wonderful! _However_, after looking through the Fascicle 16 poems today with @brookestewart, we noticed that in the Dickinson original tables (the red ones) the variations are being output as one table cell...[See here Poem 1, line 15](http://dickinson16.newtfire.org/16/1601.html).
I tried to edit the XSLT so that for the `rdg`s that contain `var`s  it output with the spans around these `rdg`s, which I got it to do!!  BUT it is now outputting these vars in _every single publication variant_ and it is still keeping the `df16` variants smushed together in a table cell with the variations...
[Here is the XSLT I used.](https://github.com/ebeshero/EmilyDickinson16/blob/master/DickinsonLineNumbers/Dickinson16FINAL.xsl)
[And here is the output that I'm getting.](http://htmlpreview.github.io/?https://github.com/ebeshero/EmilyDickinson16/blob/master/Web%20Files/16/1601Experiment.html)
You'll need to toggle all of the publications on to see what I mean...
I know I'm being a pain in the butt! :( I'm so sorry I keep asking for help from you both, but I really need it to do Dickinson's variations justice. Thanks in advance!

**Comment by ebeshero** [comment]
I'm taking a look now! @nlottig94

**Comment by nlottig94** [comment]
Thank you so much @ebeshero!!!

**Comment by nlottig94** [comment]
Any luck, @ebeshero?

**Comment by ebeshero** [comment]
I didn't actually start looking at it yet--sorry. Fell asleep after I typed that last night. I'll take a look now. There's something different between the way you're processing Fascicle 6 vs. Fascicle 16, isn't there? So, when you run your XSLT over a Fascicle **6** poem encoded in a similar way with Var01, Var02, etc. you get the output you want? If that is the case, you want to look at why Fascicle 6 is processing differently from Fascicle 16: The answer is either in your XSLT processing or in your XML source encoding. @nlottig94

**Comment by ebeshero** [comment]
@nlottig94 Can you tell me whether the XSLT you are using to process Fascicle 6 is different from the one you use to process Fascicle 16, and if they are different, can you point me to the one you're using for Fascicle 6? 

Or is this the same XSLT for both fascicles?

**Comment by ebeshero** [comment]
@nlottig94 Okay--I see they must be different, and the difference is your handling of VAR01, VAR01, etc.  This is your current XSLT for Fascicle 6, right? 
https://github.com/nlottig94/EmilyDickinson6/blob/master/Web-Files/6/xml/Dickinson6FINAL.xsl

I think what's wrong for Fascicle 16 is simply that you're processing the output of VAR01 when it appears together with other witnesses in an rdg, and you really want special behavior for it separately from the other witnesses, when you're processing views of Dickinson's ms. Let's review that and make sure we are all clear about this: What behavior do you want for Var01, Var02, and Var03? Aren't these always displayed together in the ms view, and we don't toggle these off and on?

**Comment by nlottig94** [comment]
@ebeshero
Yes, that's exactly what I'm trying to do.  If you go into the 'Web Files' folder and open 1601variantExperiment2, you can see how I want it to look (other than all of the publications are together because the CSS won't work on them anymore).

**Comment by ebeshero** [comment]
@nlottig94 I'm working on this in transit, so I'll update through the afternoon: The first thing I notice is that your website at newtfire is rendering the poem differently than the way it comes out when I view it locally within your GitHub repo: 
The website view of 1601: http://dickinson16.newtfire.org/16/1601.html isn't showing "VAR1" and "VAR2" at all--but the GitHub version, when I run it and open it locally **is** showing it, but is outputting those VAR1 and VAR2 readings inside every witness: http://htmlpreview.github.io/?https://github.com/ebeshero/EmilyDickinson16/blob/master/Web%20Files/16/1601Experiment.html

 I'm just documenting the problem so it's clear to us both. First of all, you may want to make sure that you're posting the up-to-date CSS and output HTML files on newtfire...And I think we'll figure out what's up with those VAR spans--I'll ping back here in a little while with more.

**Comment by nlottig94** [comment]
@ebeshero
I haven't pushed the ones with the variants to the site because it's not correct, so I'm just going to leave what I have alone until we figure something out.

**Comment by ebeshero** [comment]
@nlottig94 Nicole--I've been working on this most of the day and I think there's a conceptual problem with how we have to access Dickinson's vars, that maybe we need to solve by altering the encoding of the Fascicle 16 collection so it's easier to access Dickinson's variants along with the other witnesses. This is some thorny territory that I can't resolve tonight--and you shouldn't worry about it either for right now, but we'll figure it out! Possibly tomorrow...

**Comment by nlottig94** [comment]
I was thinking the same thing that we may need to change our coding style.  Thank you for your help so far!  Just so you know, @brookestewart and I are meeting tomorrow to get the final things figured out for Dickinson, so we won't have the site quite finished until maybe Saturday.  I hope this is okay!

**Comment by ebeshero** [comment]
@nlottig94 @brookestewart I think I figured out a way to handle the XSLT, [in this commit](https://github.com/ebeshero/EmilyDickinson16/commit/bf7fdb4d2fb0d9885e3bf67f0d0cc11a20920e6d)!
I used modal XSLT. Here's [a sample of the output](http://htmlpreview.github.io/?https://github.com/ebeshero/EmilyDickinson16/blob/bf7fdb4d2fb0d9885e3bf67f0d0cc11a20920e6d/Web%20Files/16/1601ebbOutput.html) 
but your CSS doesn't work. Notice, though, that we're outputting the right things in the right places--so I think if the two of you study the code, you'll figure out how to repair the CSS to fit it. 

I didn't add comments to the XSLT (yet), to explain what I did, so I'll do that later on. 

Whew!

**Comment by nlottig94** [comment]
@ebeshero 
@brookestewart and I looked at the XSLT and it is working with the CSS!!!  I understand everything that you did!! Thanks so much!!!

**Comment by ebeshero** [comment]
@nlottig94 Yay! I guess the GitHub view was running the wrong CSS when I tried it. I am glad it is working! Modal XSLT and treating df16 as a separate thing, not one of the normal witnesses, is what makes it possible. I am glad I finally figured it out--the code was tricky to rewrite!

**Comment by nlottig94** [comment]
@ebeshero 
It was because you took out the `xsl:choose` with just the `@wit` for the `rdg` elements, so it was only catching the `{$wit}text`. See what I mean? [Look at what I changed.](https://github.com/ebeshero/EmilyDickinson16/blob/master/DickinsonLineNumbers/Dickinson16FINAL.xsl).

**Comment by nlottig94** [comment]
@ebeshero 
The new Fascicle 16 poem htmls are up and running on the site!

**Comment by ebeshero** [comment]
@nlottig94 Oh good!! Yes, I deleted those conditionals when I wasn't getting output at one point and had generated a lot of commented-out chunks. So restoring those puts back the classes you need for the CSS--huzzah!!

**Comment by ebeshero** [comment]
@nlottig94 I know this means a lot of work, but how about writing up a special page to go with the site, walking the specialized coders in your audience through the steps of the XSLT as it processes a fascicle 16 poem? It would be really helpful for future projects that deal with versioning, and maybe for future people who want to help build your Dickinson project! And it would be a great way to document the hard work we did this past week so we don't forget the important parts, like how we used xsl param and how modal XSLT helps with the particular coding challenge posed by the app and rdg witness markup.
